docs/pages/guides/testing.mdx

```diff
@@ -3,6 +3,197 @@ import { ComingSoon } from '../../components/ComingSoon';
+The second is a `graphQLRequest` function, which lets you run GraphQL requests over HTTP using the [`supertest`](https://github.com/visionmedia/supertest) library
+
+x> The test runner will drop all data in your database on each run. Make sure you do not run your tests against a system with live data.
+
+## Writing tests
+
+In general you will want to run tests which check the behaviour of any custom code that you write as part of your Keystone system.
+This includes things like access control, hooks, virtual fields, and GraphQL API extensions.
+
+### Context API
+
+The [context API](../apis/context) lets you easily manipulate data in your system.
+We can use the [list items API](../apis/list-items) to ensure that we can do basic CRUD operations.
+
+```typescript
+runner(async ({ context }) => {
+  const person = await context.lists.Person.createOne({
+    data: { name: 'Alice', email: 'alice@example.com', password: 'super-secret' },
+    query: 'id name email password { isSet }',
+  });
+  expect(person.name).toEqual('Alice');
+  expect(person.email).toEqual('alice@example.com');
+  expect(person.password.isSet).toEqual(true);
+})
+```
+
+This API works well when we expect an operation to succeed.
+If we expect an operation to fail we can use the `context.graphql.raw` operation to check that both the `data` and `errors` returned by a query are what we expect.
+
+```typescript
+runner(async ({ context }) => {
+  // Create user without the required `name` field
+  const { data, errors } = await context.graphql.raw({
+    query: `mutation {
+      createPerson(data: { email: "alice@example.com", password: "super-secret" }) {
+        id name email password { isSet }
+      }
+    }`,
+  });
+  expect(data.createPerson).toBe(null);
+  expect(errors).toHaveLength(1);
+  expect(errors[0].path).toEqual(['createPerson']);
+  expect(errors[0].message).toEqual('You attempted to perform an invalid mutation');
+})
+```
+
+The `context.withSession()` function can be used to run queries as if you were logged in as a particular user.
+This can be useful for testing the behaviour of your access control rules.
+In the example below, the access control only allows users to update their own tasks.
+
+```
+runner(async ({ context }) => {
+  // Create some users
+  const [alice, bob] = await context.lists.Person.createMany({
+    data: [
+      { data: { name: 'Alice', email: 'alice@example.com', password: 'super-secret' } },
+      { data: { name: 'Bob', email: 'bob@example.com', password: 'super-secret' } },
+    ],
+  });
+
+  // Create a task assigned to Alice
+  const task = await context.lists.Task.createOne({
+    data: {
+      label: 'Experiment with Keystone',
+      priority: 'high',
+      isComplete: false,
+      assignedTo: { connect: { id: alice.id } },
+    },
+  });
+
+  // Check that we can't update the task when logged in as Bob
+  const { data, errors } = await context
+    .withSession({ itemId: bob.id, data: {} })
```

Comment:
This doesn't seem right, assuming using the auth package, it would be this, right?
```suggestion
    .withSession({ listKey: 'Person', itemId: bob.id.toString(), data: { id: bob.id.toString() } })
```